src/map.rs

```diff
@@ -344,6 +344,139 @@ impl<K: Ord, V: Val<A> + Default, A: Actor> Map<K, V, A> {
             _ => { /* we've seen all keys this clock has seen */ }
         }
     }
+
+    /// Gets an iterator over the keys of the `Map`.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use crdts::Map;
+    /// use crdts::MVReg;
+    /// use crdts::CmRDT;
+    ///
+    /// type Actor = &'static str;
+    /// type Key = &'static str;
+    ///
+    /// let actor = "actor";
+    ///
+    /// let mut map: Map<i32, MVReg<Key, Actor>, Actor> = Map::new();
+    ///
+    /// let add_ctx = map.read_ctx().derive_add_ctx(actor);
+    /// map.apply(map.update(100, add_ctx, |v, a| v.write("foo", a)));
+    ///
+    /// let add_ctx = map.read_ctx().derive_add_ctx(actor);
+    /// map.apply(map.update(50, add_ctx, |v, a| v.write("bar", a)));
+    ///
+    /// let add_ctx = map.read_ctx().derive_add_ctx(actor);
+    /// map.apply(map.update(200, add_ctx, |v, a| v.write("baz", a)));
+    ///
+    ///
+    /// let mut items = Vec::new();
+    ///
+    /// for read_ctx in map.keys() {
+    ///     items.push(*read_ctx.val);
+    /// }
+    ///
+    /// items.sort();
+    ///
+    /// assert_eq!(items, &[50, 100, 200]);
+    /// ```
+    pub fn keys(&self) -> impl IntoIterator<Item = ReadCtx<&K, A>> {
+        self.entries.keys().map(move |k| ReadCtx {
+            add_clock: self.clock.clone(),
+            rm_clock: self.clock.clone(),
+            val: k,
+        })
+    }
+
+    /// Gets an iterator over the values of the `Map`.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use crdts::Map;
+    /// use crdts::MVReg;
+    /// use crdts::CmRDT;
+    ///
+    /// type Actor = &'static str;
+    /// type Key = &'static str;
+    ///
+    /// let actor = "actor";
+    ///
+    /// let mut map: Map<i32, MVReg<Key, Actor>, Actor> = Map::new();
+    ///
+    /// let add_ctx = map.read_ctx().derive_add_ctx(actor);
+    /// map.apply(map.update(100, add_ctx, |v, a| v.write("foo", a)));
+    ///
+    /// let add_ctx = map.read_ctx().derive_add_ctx(actor);
+    /// map.apply(map.update(50, add_ctx, |v, a| v.write("bar", a)));
+    ///
+    /// let add_ctx = map.read_ctx().derive_add_ctx(actor);
+    /// map.apply(map.update(200, add_ctx, |v, a| v.write("baz", a)));
+    ///
+    ///
+    /// let mut items = Vec::new();
+    ///
+    /// for read_ctx in map.values() {
+    ///     items.push(read_ctx.val.read().val[0]);
+    /// }
+    ///
+    /// items.sort();
+    ///
+    /// assert_eq!(items, &["bar", "baz", "foo"]);
+    /// ```
+    pub fn values(&self) -> impl IntoIterator<Item = ReadCtx<&V, A>> {
+        self.entries.values().map(move |v| ReadCtx {
+            add_clock: self.clock.clone(),
+            rm_clock: self.clock.clone(),
```

Comment:
Not sure if this is meant to be self.clock or the clock of the element, like in `get`?

Reply:
Yeah you got it right in the next commit, we need the entry clock